### PR TITLE
Initial Implementation

### PR DESCRIPTION
**Pull request by slax57**

### CODE REVIEW

File: packages/ra-auth-msal/Readme.md [hunk 1]
```diff
@@ -0,0 +1,378 @@
+# ra-auth-msal
+
+An auth provider for [react-admin](https://github.com/marmelab/react-admin) which handles authentication using the [Microsoft Authentication Library (MSAL)](https://learn.microsoft.com/fr-fr/azure/active-directory/develop/msal-overview).
```
Comment (djhi) [inline packages/ra-auth-msal/Readme.md:3]:
```suggestion
An auth provider for [react-admin](https://github.com/marmelab/react-admin) that handles authentication using the [Microsoft Authentication Library (MSAL)](https://learn.microsoft.com/fr-fr/azure/active-directory/develop/msal-overview).
```

File: packages/ra-auth-msal/Readme.md [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+This is useful for instance if you want to use [Azure Active Directory](https://azure.microsoft.com/en-us/services/active-directory/) to authenticate your users.
```
Comment (djhi) [inline packages/ra-auth-msal/Readme.md:5]:
```suggestion
This is useful when using [Azure Active Directory](https://azure.microsoft.com/en-us/services/active-directory/) to authenticate your users.
```

File: packages/ra-auth-msal/Readme.md [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        >
+            <Resource name="posts" {...posts} />
+      </Admin>
+    </BrowserRouter>
+   );
+};
+export default App;
+```
+
+**Tip:** You need to wrap your `<Admin>` component in a `<BrowserRouter>` for this library to work. Indeed, MSAL uses a hash-based routing strategy when redirecting back to your app, which would be incompatible with a `<HashRouter>`.
```
Comment (djhi) [inline packages/ra-auth-msal/Readme.md:83]:
```suggestion
**Tip:** You need to wrap your `<Admin>` component in a `<BrowserRouter>` for this library to work. Indeed, MSAL uses a hash-based routing strategy when redirecting back to your app, which is incompatible with a `<HashRouter>`.
```

File: packages/ra-auth-msal/src/authProvider.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+}: MsalAuthProviderParams): AuthProvider => {
+  // We need to set up the redirect handler at a global scope to make sure all redirects are handled,
+  // otherwise the lib can lock up because a redirect is still marked as pending and has not been handled.
+  // Besides, we can call this handler again later and still gather the response because it is cached internally.
+  msalInstance.handleRedirectPromise();
+
+  return {
+    async login() {
+      // should never be called
+      throw Error("Not implemented");
```
Comment (djhi) [inline packages/ra-auth-msal/src/authProvider.ts:98]: It would be nice to allow users to choose whether to redirect immediatly to the third party login on checkAuth or to to do it from a user action. See https://github.com/marmelab/ra-auth-auth0/blob/main/packages/ra-auth-auth0/src/authProvider.ts

**Review by djhi — CHANGES_REQUESTED** [review]

File: packages/ra-auth-msal/src/authProvider.ts (hunk 4 above)
Comment (slax57) [reply packages/ra-auth-msal/src/authProvider.ts:98]: Right !
Besides, I forgot that we added a [demo-fake-api](https://github.com/marmelab/ra-auth-auth0/tree/main/packages/demo-fake-api) package, which will allow me to demonstrate the `httpClient` feature. I'll add this as well.
Back to WIP.

Comment (slax57) [reply packages/ra-auth-msal/src/authProvider.ts]: Done, back to RFR!

File: makefile [hunk 5]
```diff
@@ -16,8 +16,14 @@ build-demo-react-admin:
...
 
-start: ## Start the demo
+start-demo: ## Start the demo
 	@cd ./packages/demo-react-admin && yarn start
 
+start-fake-api: ## Start the fake API
+	@cd ./packages/demo-fake-api && yarn start
+
+start: ## Start the demo
+	@(trap 'kill 0' INT; ${MAKE} start-fake-api & ${MAKE} start-demo)
```
Comment (djhi) [inline makefile:26]: For my info, what is this voodoo?

Comment (slax57) [reply makefile:26]: I wanted a way to run both the fake-api and the demo in a single comment, but also kill both processes with a CTRL-C.
Maybe there's an easier way to do it, but I found this one here: https://stackoverflow.com/a/52033580/3975522 :grin: 
What do you think? Can we keep it?

Comment (djhi) [reply makefile:26]: That's nice!

File: README.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+```ts
+/**
+ * Customize this map to match the ids of the groups you created in Azure AD
+ */
+const rolesPermissionMap = {
+  "199a772f-5fbe-42ce-8adb-2838a9b3d37f": "user",
+  "e3f75483-90e9-430a-86eb-fe97ed36b941": "admin",
+};
+```
```
Comment (slax57) [inline README.md:90]: Thanks for the doc updates, indeed I missed a few steps!!

However, why did you change this part? These UUIDs are client dependent, so we should keep the dummy values in the docs to show that these need to be updated with their actual values.

File: README.md [hunk 7]
```diff
@@ -80,8 +84,8 @@ Lastly, we need to edit the `rolesPermissionMap` inside `packages/demo-react-adm
  * Customize this map to match the ids of the groups you created in Azure AD
  */
 const rolesPermissionMap = {
-  "12345678-1234-1234-1234-123456789012": "user",
-  "12345678-1234-1234-1234-123456789013": "admin",
+  "199a772f-5fbe-42ce-8adb-2838a9b3d37f": "user",
+  "e3f75483-90e9-430a-86eb-fe97ed36b941": "admin",
```
Comment (djhi) [inline README.md:88]:
```suggestion
  "12345678-1234-1234-1234-123456789012": "user",
  "12345678-1234-1234-1234-123456789013": "admin",
```

File: README.md (hunk 6 above)
Comment (slax57) [reply README.md]: Fixed by df5e0e1487968a1583706e3b5990a835fc25dd2f

**Review by djhi — APPROVED** [review]
:muscle:
